tools/firmware.toit

```diff
@@ -0,0 +1,600 @@
+    // less padding somehow.
+    padded := round_up bits.size 64 * 1024
+    bits = bits + (ByteArray padded - bits.size)
+    // We look for the last DROM segment, because it will grow into
+    // unused virtual memory, so we can extend that without relocating
+    // other segments (which we don't know how to).
+    drom := find_last_drom_segment_
+    if not drom: throw "Cannot append to non-existing DROM segment"
+    // Run through all the segments and extend the
+    // segment we just found. All segments following
+    // that one needs to be displaced in flash.
+    displacement := null
+    segments_.size.repeat:
+      segment/Esp32BinarySegment := segments_[it]
+      if segment == drom:
+        segments_[it] = Esp32BinarySegment (segment.bits + bits)
+            --offset=segment.offset
+            --address=segment.address
+        displacement = bits.size
+      else if displacement:
+        segments_[it] = Esp32BinarySegment segment.bits
+            --offset=segment.offset + displacement
+            --address=segment.address
+
+  find_last_drom_segment_ -> Esp32BinarySegment?:
+    last := null
+    segments_.do: | segment/Esp32BinarySegment |
+      address := segment.address
+      if not DROM_MAP_START <= address < DROM_MAP_END: continue.do
+      if not last or address > last.address: last = segment
+    return last
+
+  static read_segment_ bits/ByteArray offset/int -> Esp32BinarySegment:
+    address := LITTLE_ENDIAN.uint32 bits
+        offset + Esp32BinarySegment.LOAD_ADDRESS_OFFSET_
+    size := LITTLE_ENDIAN.uint32 bits
+        offset + Esp32BinarySegment.DATA_LENGTH_OFFSET_
+    start := offset + Esp32BinarySegment.HEADER_SIZE_
+    return Esp32BinarySegment bits[start..start + size]
+        --offset=offset
+        --address=address
+
+  static write_segment_ bits/ByteArray segment/Esp32BinarySegment -> none:
+    offset := segment.offset
+    LITTLE_ENDIAN.put_uint32 bits
+        offset + Esp32BinarySegment.LOAD_ADDRESS_OFFSET_
+        segment.address
+    LITTLE_ENDIAN.put_uint32 bits
+        offset + Esp32BinarySegment.DATA_LENGTH_OFFSET_
+        segment.size
+    bits.replace (offset + Esp32BinarySegment.HEADER_SIZE_) segment.bits
+
+class Esp32BinarySegment:
+  static LOAD_ADDRESS_OFFSET_ ::= 0
+  static DATA_LENGTH_OFFSET_  ::= 4
+  static HEADER_SIZE_         ::= 8
+
+  bits/ByteArray
+  offset/int
+  address/int
+
+  constructor .bits --.offset --.address:
+
+  size -> int:
+    return bits.size
+
+  end -> int:
+    return offset + HEADER_SIZE_ + size
+
+  xor_checksum -> int:
+    result := 0
+    bits.do: result ^= it
+    return result
+
+  stringify -> string:
+    return "len 0x$(%05x size) load 0x$(%08x address) file_offs 0x$(%08x offset)"
+
+// These two are the current offsets of the config data in the system image.
+// We could auto-detect them from the bin file, but they are only used files
+// from the current SDK so there's no need.
+IMAGE_DATA_SIZE ::= 1024
+IMAGE_DATA_OFFSET ::= 296
```

Comment:
These aren't used anymore, and it seems like the comment is stale too.

Reply:
Fixed.